esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+    if (url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: missing url property", name);
+    if (strncmp(url, gwLocalSecretPrefix, gwLocalSecretPrefixLength))
+        throw makeStringExceptionV(-1, "gateway %s: expected '%s...'; got '%s'", name, gwLocalSecretPrefix, url.str());
+    const char* identification = url.str() + gwLocalSecretPrefixLength;
+    StringArray tokens;
+    tokens.appendList(identification, ":", true);
+    Owned<IPTree> secret;
+    switch (tokens.ordinality())
+    {
+    case 1:
+        secret.setown(getSecret("esp", tokens.item(0)));
+        break;
+    case 2:
+        secret.setown(getVaultSecret("esp", tokens.item(0), tokens.item(1)));
+        break;
+    default:
+        throw makeStringExceptionV(-1, "gateway %s: '%s' is not of the form \"[ vault-id ':' ] secret-name\"", name, identification);
+    }
+    if (!secret)
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not identify an 'esp' category secret", name, identification);
+    if (!isEmptyString(permission) && !secret->getPropBool(permission))
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not grant '%s' permission", name, identification, permission);
+    if (!secret->getProp("url", url.clear()) || url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: '%s' missing required 'url' property; credential-only secrets not supported", name, identification);
+    const char* username = secret->queryProp("username");
+    if (isEmptyString(username) && !secret->getPropBool("insecure"))
+        throw makeStringExceptionV(-1, "gateway %s: '%s' missing expected 'username' property; set 'insecure` property to 'true' if credentials are not required", name, identification);
+    const char* password = secret->queryProp("password");
+    if (isEmptyString(username) && password)
+        throw makeStringExceptionV(-1, "gateaay %s: '%s' invalid use of password without username", name, identification);
+    adjustURL(url, username, password);
+    gateway.setProp("@url", url);
+}
+
+void EsdlServiceImpl::resolveGatewayInlineURL(IPTree& gateway) const
+{
+    const char* name = gateway.queryProp("@name");
+    StringBuffer url(gateway.queryProp("@url"));
+    if (url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: missing  required '@url' property", name);
+    const char* username = gateway.queryProp("@username");
+    const char* password = gateway.queryProp("@password");
+    if (isEmptyString(username) && password)
+        throw makeStringExceptionV(-1, "gateway %s: invalid credentials - password without username", name);
+    StringBuffer decryptedPassword;
+    if (!isEmptyString(password))
+    {
+        decrypt(decryptedPassword, password);
+        password = decryptedPassword;
+    }
+    adjustURL(url, username, password);
+    gateway.setProp("@url", url);
+}
+
+void EsdlServiceImpl::adjustURL(StringBuffer& url, const char* username, const char* password) const
```

Comment:
Can we put something in the method name about adding credentials?  adjustURL is a bit ambiguous.